cmd/controller/app/start.go

```diff
@@ -17,85 +17,205 @@ limitations under the License.
+		// configuration via versioned configuration files and flag values.
+		// Setting DisableFlagParsing=true prevents Cobra from interfering with flag parsing
+		// at all, and instead we handle it all in the RunE below.
+		DisableFlagParsing: true,
+		Run: func(cmd *cobra.Command, args []string) {
+			// initial flag parse, since we disable cobra's flag parsing
+			if err := cleanFlagSet.Parse(args); err != nil {
+				log.Error(err, "Failed to parse controller flag")
+				cmd.Usage()
+				os.Exit(1)
+			}
 
-		RunE: func(cmd *cobra.Command, args []string) error {
-			if err := o.Validate(args); err != nil {
-				return fmt.Errorf("error validating options: %s", err)
+			// check if there are non-flag arguments in the command line
+			cmds := cleanFlagSet.Args()
+			if len(cmds) > 0 {
+				log.Error(nil, "Unknown command", "command", cmds[0])
+				cmd.Usage()
+				os.Exit(1)
 			}
 
-			logf.Log.V(logf.InfoLevel).Info("starting controller", "version", util.AppVersion, "git-commit", util.AppGitCommit)
-			if err := o.RunCertManagerController(stopCh); err != nil {
-				cmd.SilenceUsage = true // Don't display usage information when exiting because of an error
-				return err
+			// short-circuit on help
+			help, err := cleanFlagSet.GetBool("help")
+			if err != nil {
+				log.Info(`"help" flag is non-bool, programmer error, please correct`)
+				os.Exit(1)
+			}
+			if help {
+				cmd.Help()
 			}
 
-			return nil
+			// set feature gates from initial flags-based config
+			if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(controllerConfig.FeatureGates); err != nil {
+				log.Error(err, "Failed to set feature gates from initial flags-based config")
+				os.Exit(1)
+			}
+
+			if err := options.ValidateControllerFlags(controllerFlags); err != nil {
+				log.Error(err, "Failed to validate controller flags")
+				os.Exit(1)
+			}
+
+			if configFile := controllerFlags.Config; len(configFile) > 0 {
+				controllerConfig, err = loadConfigFile(configFile)
+				if err != nil {
+					log.Error(err, "Failed to load controller config file", "path", configFile)
+					os.Exit(1)
+				}
+
+				if err := controllerConfigFlagPrecedence(controllerConfig, args); err != nil {
+					log.Error(err, "Failed to merge flags with config file values")
+					os.Exit(1)
+				}
+				// update feature gates based on new config
+				if err := utilfeature.DefaultMutableFeatureGate.SetFromMap(controllerConfig.FeatureGates); err != nil {
+					log.Error(err, "Failed to set feature gates from config file")
+					os.Exit(1)
+				}
+			}
+
+			// Start the controller
+			if err := Run(controllerConfig, stopCh); err != nil {
+				log.Error(err, "Failed to run the controller")
+				os.Exit(1)
+			}
 		},
-		SilenceErrors: true, // Errors are already logged when calling cmd.Execute()
 	}
 
-	flags := cmd.Flags()
-	o.ControllerOptions.AddFlags(flags)
-	utilfeature.DefaultMutableFeatureGate.AddFlag(flags)
```

Comment:
I think we've lost the `--feature-gates` flag by removing this line (see `go run cmd/controller/main.go --help`) and it needs re-instating. Hopefully should work in the same way. (Webhook in this case is not a good example as the feature flags there were implemented a bit differently https://github.com/cert-manager/cert-manager/pull/4670

Reply:
Aww, good catch. I'll add this back in.